models/upload_sessions_test.go

```diff
@@ -295,3 +296,94 @@ func (ms *ModelSuite) Test_GetTreasureIndexes() {
 	// When we update MergeIndexes to hash the indexes, this test will start failing
 	ms.Equal(expectedIndexes, actualIndexes)
 }
+
+func (ms *ModelSuite) Test_CalculatePayment_Less_Than_1_GB() {
+
+	currentStoragePeg := models.StoragePeg
+
+	defer func() { models.StoragePeg = currentStoragePeg }()
+
+	fileSizeBytes := 9000000
+
+	models.StoragePeg = 64
+	storageLengthInYears := 3
+
+	u := models.UploadSession{
+		Type:                 models.SessionTypeAlpha,
+		GenesisHash:          "genHash",
+		NumChunks:            2,
+		FileSizeBytes:        fileSizeBytes,
+		StorageLengthInYears: storageLengthInYears,
+	}
+
+	vErr, err := u.StartUploadSession()
+	invoice := u.GetInvoice()
+	ms.Nil(err)
+	ms.Equal(0, len(vErr.Errors))
+
+	// expecting to be charged for 1 full sector even though we aren't using the whole sector
+	ms.Equal(decimal.NewFromFloatWithExponent(0.04687500000000001, -16), invoice.Cost)
+}
+
+func (ms *ModelSuite) Test_CalculatePayment_Greater_Than_1_GB() {
+
+	defer oyster_utils.SetBrokerMode(oyster_utils.ProdMode)
+	oyster_utils.SetBrokerMode(oyster_utils.TestModeDummyTreasure)
+
+	currentStoragePeg := models.StoragePeg
+
+	defer func() { models.StoragePeg = currentStoragePeg }()
+
+	fileSizeBytes := 1500000000
+
+	models.StoragePeg = 64
+	storageLengthInYears := 3
+
+	u := models.UploadSession{
+		Type:                 models.SessionTypeAlpha,
+		GenesisHash:          "genHash",
+		NumChunks:            2,
+		FileSizeBytes:        fileSizeBytes,
+		StorageLengthInYears: storageLengthInYears,
+	}
+
+	vErr, err := u.StartUploadSession()
+	invoice := u.GetInvoice()
+	ms.Nil(err)
+	ms.Equal(0, len(vErr.Errors))
+
+	// expecting to be charged for 2 full sectors even though we're only using 1.5
+	ms.Equal(decimal.NewFromFloatWithExponent(0.09375000000000001, -16), invoice.Cost)
+}
+
+func (ms *ModelSuite) Test_CalculatePayment_2_GB() {
+
+	defer oyster_utils.SetBrokerMode(oyster_utils.ProdMode)
+	oyster_utils.SetBrokerMode(oyster_utils.TestModeDummyTreasure)
+
+	currentStoragePeg := models.StoragePeg
+
+	defer func() { models.StoragePeg = currentStoragePeg }()
+
+	fileSizeBytes := 2000000000
+
+	models.StoragePeg = 64
+	storageLengthInYears := 3
+
+	u := models.UploadSession{
+		Type:                 models.SessionTypeAlpha,
+		GenesisHash:          "genHash",
+		NumChunks:            2,
+		FileSizeBytes:        fileSizeBytes,
+		StorageLengthInYears: storageLengthInYears,
+	}
+
+	vErr, err := u.StartUploadSession()
+	invoice := u.GetInvoice()
+	ms.Nil(err)
+	ms.Equal(0, len(vErr.Errors))
+
+	// expecting to be charged for 2 full sectors
+	// should be same price as previous test
+	ms.Equal(decimal.NewFromFloatWithExponent(0.09375000000000001, -16), invoice.Cost)
```

Comment:
The 2nd parameter in NewFloatWithExponent is the exponent, and if you give it `decimal.NewFromFloatWithExponent(0.09375, -16)` it ignores the -16 and the exponent will be -5.  And when you do `decimal.NewFromFloatWithExponent(0.0937500000000000, -16)` it truncates back to -5.  Hence the additional 1, which gets cut off but it makes the result have the right exp value.